docs/FlowClassifyHook.md

```diff
@@ -0,0 +1,302 @@
+- Multiple eBPF programs can be invoked for a single stream, but the decision is finalized based on the first conclusive return code.
+- No stream mutation is supported through these hooks (only inspection and allow/block).
+
+### WFP Implementation Details
+
+The stream inspection extension integrates with the Windows Filtering Platform (WFP)
+at the flow_established layer to indicate flows for classification and at the stream layer to
+classify flows.
+
+Currently only inline stream classification will be supported.
+
+#### WFP Action Mapping
+
+The hook implementation maps eBPF program return values to specific WFP actions:
+
+- **FLOW_CLASSIFY_ALLOW**: Returns `FWP_ACTION_PERMIT` and deletes the WFP flow context. This allows the current segment and terminates further classification for this flow by removing the flow context.
+
+- **FLOW_CLASSIFY_BLOCK**: Returns `FWP_ACTION_BLOCK` with the `FWPS_CLASSIFY_OUT_FLAG_ABSORB` flag set. This blocks the current segment and terminates the connection by dropping all subsequent packets.
+
+- **FLOW_CLASSIFY_NEED_MORE_DATA**: Returns `FWP_ACTION_PERMIT`. This allows the current segment to proceed while maintaining the flow context for continued classification of subsequent segments.
+
+#### Stream Inspection Model
+
+The implementation supports **inline stream inspection only**, meaning:
+
+- All classification decisions are made in the WFP callout
+  - No out-of-band or deferred processing is supported
+- Stream data is processed segment-by-segment as it arrives
+- No reassembly or reordering of TCP segments is performed by the hook (this is handled by the TCP stack before reaching the stream layer).
+- Data is allowed through when `FLOW_CLASSIFY_NEED_MORE_DATA` is returned.
+  - The allow allow/block decision is at the flow level (not segment level).
+  - **Note**: The `countBytesEnforced` field is not currently used in this implementation as the hook focuses on connection-level allow/block decisions rather than byte-level enforcement.
+
+#### Stream Data Handling
+
+The hook provides direct access to stream segment data through the context's `data_start` and `data_end` pointers:
+
+- **Contiguous Data**: When possible, pointers reference the original NET_BUFFER data directly for zero-copy access
+- **Non-contiguous Data**: For fragmented NET_BUFFERs, data is copied into a temporary buffer using `FwpsCopyStreamDataToBuffer0` to provide contiguous access to eBPF programs
+- **Data Lifetime**: Stream data pointers are only valid during the program invocation and must not be accessed after the program returns
+
+#### Resource Management
+
+The implementation manages WFP resources efficiently:
+
+- **Conditional Callouts**: Stream layer callouts use the `FWP_CALLOUT_FLAG_CONDITIONAL_ON_FLOW` flag to ensure they are only invoked for flows requiring classification
+- **Context Cleanup**: Flow contexts are automatically cleaned up when flows are allowed, blocked, or naturally terminated
```

Comment:
For both context cleanup and memory management - can you update the text slightly, to mention that the stream extension does the following?

Alternatively, if it is WFP that does something, please call that out here too. Sometimes, WFP callouts hand memory off to WFP platform, which is responsible for cleaning it up later, so it is good to be precise here about what we are responsible for/doing w.r.t. memory.)